### PR TITLE
Fix #5488: change default for logging to `+nowrap`

### LINKED ISSUE #5488: RFC: change default for logging to `+nowrap`
I'm curious what others think about this, hence I'm putting this up for discussion. And being biased, I'm also specifically interested in hearing about reasons why `+nowrap` is undesirable.

By default, cabal line-wraps its logging output to hardcoded line-length of 79 characters via `wrapText`:

https://github.com/haskell/cabal/blob/3b390592f239bab3545da6226e3730b8ffab3731/Cabal/Distribution/Utils/Generic.hs#L108-L131

In #4091, a verbosity level modified `+nowrap` was introduced which suppresses this line-wrapping;

IMO, having `cabal` pre-linewrap its output interacts badly with terminals or editors if these re-wrap dynamically when resizing the window, readability is IMHO not improved by line-wrapping, and it makes it harder to copy and paste as the terminal can't easily reconstruct the original log message's line endings, and thus the pre-wrapped text is copy-n-pasted into the target including the line-wrapping (which may again not be adequate for the target text-body width).

As an example, this here is a solver failure pre-wrapped via `wrapText`

```
$ cabal new-repl -w ghc-8.6.1 --build-depends 'lens'
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: fake-package-0 (user goal)
[__1] trying: lens-4.17 (dependency of fake-package)
[__2] trying: transformers-0.5.5.0/installed-0.5... (dependency of lens)
[__3] next goal: hashable (dependency of lens)
[__3] rejecting: hashable-1.2.7.0 (conflict: transformers =>
base==4.12.0.0/installed-4.1..., hashable => base>=4.4 && <4.12)
[__3] rejecting: hashable-1.2.6.1 (conflict: transformers =>
base==4.12.0.0/installed-4.1..., hashable => base>=4.4 && <4.11)
[__3] rejecting: hashable-1.2.6.0 (conflict: transformers =>
base==4.12.0.0/installed-4.1..., hashable => base>=4.4 && <4.10)
[__3] rejecting: hashable-1.2.5.0, hashable-1.2.4.0 (conflict: transformers =>
base==4.12.0.0/installed-4.1..., hashable => base>=4.0 && <4.10)
[__3] rejecting: hashable-1.2.3.3, hashable-1.2.3.2 (conflict: transformers =>
base==4.12.0.0/installed-4.1..., hashable => base>=4.0 && <4.9)
[__3] rejecting: hashable-1.2.3.1, hashable-1.2.3.0, hashable-1.2.2.0,
hashable-1.2.1.0, hashable-1.2.0.10, hashable-1.2.0.9, hashable-1.2.0.8,
hashable-1.2.0.7, hashable-1.2.0.6, hashable-1.2.0.5, hashable-1.2.0.4,
hashable-1.2.0.3, hashable-1.2.0.2, hashable-1.2.0.1, hashable-1.2.0.0,
hashable-1.1.2.5, hashable-1.1.2.4, hashable-1.1.2.3 (conflict: transformers
=> base==4.12.0.0/installed-4.1..., hashable => base>=4.0 && <4.8)
[__3] rejecting: hashable-1.1.2.2, hashable-1.1.2.1, hashable-1.1.2.0,
hashable-1.1.1.0, hashable-1.1.0.0, hashable-1.0.1.1, hashable-1.0.1.0,
hashable-1.0.0 (conflict: lens => hashable>=1.1.2.3 && <1.3)
[__3] fail (backjumping, conflict set: hashable, lens, transformers)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: hashable, lens, base, transformers,
fake-package
```

...and the same with `+nowrap`:

```

$ cabal new-repl -w ghc-8.6.1 --build-depends 'lens' -vnormal+nowrap
Resolving dependencies...
cabal: Could not resolve dependencies:
[__0] trying: fake-package-0 (user goal)
[__1] trying: lens-4.17 (dependency of fake-package)
[__2] trying: transformers-0.5.5.0/installed-0.5... (dependency of lens)
[__3] next goal: hashable (dependency of lens)
[__3] rejecting: hashable-1.2.7.0 (conflict: transformers => base==4.12.0.0/installed-4.1..., hashable => base>=4.4 && <4.12)
[__3] rejecting: hashable-1.2.6.1 (conflict: transformers => base==4.12.0.0/installed-4.1..., hashable => base>=4.4 && <4.11)
[__3] rejecting: hashable-1.2.6.0 (conflict: transformers => base==4.12.0.0/installed-4.1..., hashable => base>=4.4 && <4.10)
[__3] rejecting: hashable-1.2.5.0, hashable-1.2.4.0 (conflict: transformers => base==4.12.0.0/installed-4.1..., hashable => base>=4.0 && <4.10)
[__3] rejecting: hashable-1.2.3.3, hashable-1.2.3.2 (conflict: transformers => base==4.12.0.0/installed-4.1..., hashable => base>=4.0 && <4.9)
[__3] rejecting: hashable-1.2.3.1, hashable-1.2.3.0, hashable-1.2.2.0, hashable-1.2.1.0, hashable-1.2.0.10, hashable-1.2.0.9, hashable-1.2.0.8, hashable-1.2.0.7, hashable-1.2.0.6, hashable-1.2.0.5, hashable-1.2.0.4, hashable-1.2.0.3, hashable-1.2.0.2, hashable-1.2.0.1, hashable-1.2.0.0, hashable-1.1.2.5, hashable-1.1.2.4, hashable-1.1.2.3 (conflict: transformers => base==4.12.0.0/installed-4.1..., hashable => base>=4.0 && <4.8)
[__3] rejecting: hashable-1.1.2.2, hashable-1.1.2.1, hashable-1.1.2.0, hashable-1.1.1.0, hashable-1.1.0.0, hashable-1.0.1.1, hashable-1.0.1.0, hashable-1.0.0 (conflict: lens => hashable>=1.1.2.3 && <1.3)
[__3] fail (backjumping, conflict set: hashable, lens, transformers)
After searching the rest of the dependency tree exhaustively, these were the goals I've had most trouble fulfilling: hashable, lens, base, transformers, fake-package
```

Personally, I consider the `+nowrap` version easier to parse visually, as it retains more of its original structure. I don't have to cognitively unwrap long lines into their original form to unpack the information.

...opinions?

### PR DESCRIPTION
**Pull request by yvan-sraka**
This PR modifies `cabal` behavior by change default for logging to `+nowrap` and so it introduces a `+wrap` modifier :)

**Checklist:**

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.
* [ ] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

Bonus points for added automated tests!

### CODE REVIEW

**Comment by Kleidukos** [comment]
@yvan-sraka good initiative! Would you mind showing us what it looks like before and after? :)

**Comment by yvan-sraka** [comment]
> @yvan-sraka good initiative! Would you mind showing us what it looks like before and after? :)

Yes! It's highlighted in the linked issue https://github.com/haskell/cabal/issues/5488

**Review by Kleidukos — APPROVED** [review]

**Comment by andreabedini** [comment]
:thinking: this does not just flip a default setting right?

If the default is `-vnormal+nowrap`, how does one get the old behaviour? `-vnormal+wrap` does not seem to work.

Also, I am not sure adding a new constructor to `VerbosityFlag` is the right thing to do:
```
ghci> normal
Verbosity {vLevel = Normal, vFlags = fromList [], vQuiet = False}
ghci> verboseWrap normal
Verbosity {vLevel = Normal, vFlags = fromList [VWrap], vQuiet = False}
ghci> verboseNoWrap $ verboseWrap normal
Verbosity {vLevel = Normal, vFlags = fromList [VNoWrap,VWrap], vQuiet = False}
```

What would this last verbosity mean?

**Comment by yvan-sraka** [comment]
Thanks, @andreabedini! I updated the PR (removing the `VNoWrap` value from the `VerbosityFlag` enum), it, as expected, does not wrap the output by default now. But I can't figure out why `-vnormal+wrap` flag does not reproduce the current Cabal default behavior (wrapping the output) … any clue?

Also, since I couldn't test it, it will be only by guessing … But I believe the final behavior will be that, e.g., `-vnormal+wrap+nowrap` will make the output unwrapped while `-vnormal+nowrap+wrap` will wrap it, if I understood correctly the parser logic here!

**Comment by yvan-sraka** [comment]
I'll try to rephrase and better explain my bug, so I will find help in solving it:

The TLDR of https://github.com/haskell/cabal/pull/9160#issuecomment-1798232254 is that I realized that after my changes that now the output is unwrapped by default (what I want) but I can't wrap it anymore (with a verbosity modifier) which is a bug ...

The commands I'm using to experiment are to compare the output of `cabal build -vnormal+unwrap` (the `cabal-install` version provided by Nix) vs. the one of `cabal run cabal-install -- build -vnormal+wrap` (the `cabal-binaries` I just build and that contains my current changes).

But last time I ran my tests, I got `Error: cabal: invalid argument to option '-v': normal+wrap` on `/nix/store/fykw1mb27xz82jj0530w945kqihanrw4-cabal-install-3.10.1.0/bin/cabal` which confuse me a lot ... can someone help me find the right command invocations to test my changes?

After that, to sum up what we discover with @Kleidukos, is that https://github.com/haskell/cabal/blob/master/Cabal/src/Distribution/Simple/Utils.hs#L662 seems to never be called (while https://github.com/haskell/cabal/blob/master/Cabal/src/Distribution/Verbosity.hs#L202 is called), and I don't understand why ... maybe some hint on how to use the Cabal REPL to debug that (is there a documentation somewhere on that?) could help! 🙂

Thanks in advance to anyone that could help me better understand what happens here!
